VS Plugin - Adding new backtest ToolWindow

Martin-Molinero opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
<!--- Describe your changes in detail -->
This PR creates a new Visual Studio tool window that allows a user to:
- select working project
- see current/previous/on going backtests, for the working project, and their result status
- open and delete backtests

#### Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/QuantConnect/Lean/issues/1842

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Improve VS plugin value

#### Requires Documentation Change
<!--- Please indicate if these changes will require updates to documentation, and if so, specify what changes are required -->
Yes. I'm changing the VS plugin documentation.

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manually VS15 and VS17
#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

mchandschuh commented on VisualStudioPlugin/IBacktestObserver.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        void NewBacktest(int projectId, Api.Backtest backtestStatus);
+        void UpdateBacktest(int projectId, Api.Backtest backtestStatus);
+        void BacktestFinished(int projectId, Api.Backtest backtestStatus);
```
this may be me being nit picky :D

I love the 'eventy' sound of `BacktestFinished` -- it matches well with 'observer' -- what are your thoughts on renaming `NewBacktest` to something like `BacktestCreated` or `BacktestStarted`, and likewise for `UpdateBacktest` to `BacktestStatusUpdated`?

Another options is simply `Started`, `Updated`, and `Finished` since the 'backest' portion is already implied via the type name and the parameter of type `Api.Backtest`.

Martin-Molinero replied:
Your right! good suggestion, I'll use Backtest{event} format 👍

mchandschuh commented on VisualStudioPlugin/SolutionExplorerMenuCommand.cs:
```diff
@@ -53,6 +53,11 @@ internal sealed class SolutionExplorerMenuCommand
+        /// <summary>
+        /// Observer for the status of backtests launched from the plugin
+        /// </summary>
+        private IBacktestObserver _backtestObserver;
```
should this field be marked as read-only?

Martin-Molinero replied:
yes, good catch!

mchandschuh commented on VisualStudioPlugin/readme.md:
```diff
@@ -68,6 +70,19 @@ This feature allows the user to upload to QC one or more files to a target proje
+> ![alt tag](Resources/toolWindow.PNG)
+
+##### How To Open
+- From the Visual Studio IDE go to ‘View’ menu → ‘Other Windows’ → ‘QuantConnect’
```
Thanks for keeping up with the documentation!

Martin-Molinero commented on VisualStudioPlugin/SolutionExplorerMenuCommand.cs:
```diff
@@ -262,20 +267,35 @@ private bool UploadFilesToServer(int selectedProjectId, IEnumerable<SelectedItem
+                backtestStatus = await System.Threading.Tasks.Task.Run(() =>
+                {
+                    Thread.Sleep(5000);
+                    return api.ReadBacktest(projectId, backtestId);
```
your right. The suggested code works great, and using Delay seems to make the algorithm run faster!

Martin-Molinero commented:
I've implemented the changes requested by @mchandschuh . Also I've done the following changes:
 - Add missing try catch. Required for network or Api issues
 - I've added the option for AuthenticationCommand::Login() not to show login dialog which blocks the UI. This change solves the issue bellow. Now when created, the tool window will not show login dialog.
	> Given endpoint has no internet access, QC tool windows is open, close VS, open VS, login dialog opens and blocks solution load and for some VS bug all icons break too
 - Moved 'api.Connected' to a new task to avoid blocking UI
 - Some minor changes to documentation
 - Now when Login dialog is closed it will shown a message in the status bar.

mchandschuh commented on VisualStudioPlugin/SolutionExplorerMenuCommand.cs:
```diff
@@ -239,7 +258,7 @@ private bool UploadFilesToServer(int selectedProjectId, IEnumerable<SelectedItem
             while (compileStatus.State == Api.CompileState.InQueue)
             {
-                Thread.Sleep(2000);
+                System.Threading.Tasks.Task.Delay(2000);
```
Since this task is not awaited, this method will keep on going. I'd be interested to see how many times we invoke `api.ReadCompile`  -- typically async-ness infects the entire call stack (or until someone does a `Task.Run` or similar).

Martin-Molinero replied:
Your right! the method kept on going 😅, it queried a couple of times (in debug).
I've changed it to work with the same async logic as the backtest method does 👍 . With the change, in debug it's also queering twice.

mchandschuh replied:
Very cool, thanks for updating. I think this is ready for merge -- we may want to browse the vs plugin project for any other usages of `Thread.Sleep` and see if w can unroll them into `async/await` like we've done here.  If it becomes common, an `async` version of the API methods might be useful and would remove all the `Task.Run` calls and make it *even faster* :)

Martin-Molinero replied:
I've doubled checked and there are no other `Thread.Sleep` in the plugin
